### PR TITLE
Add `trust_remote_code` to trainer configs

### PR DESCRIPTION
**Pull request by qgallouedec**
Adds a top-level `trust_remote_code: bool = False` field to the trainer configs

A regression test (`test_trust_remote_code`) is added to each trainer test class. It asserts that the trainer refuses to load `trl-internal-testing/tiny-RemoteForCausalLM` without the flag and succeeds with it.

### Why not just rely on `model_init_kwargs`?

`model_init_kwargs["trust_remote_code"] = True` already works for the model load, but it does not cover  the processor / tokenizer load (`AutoProcessor.from_pretrained` / `AutoTokenizer.from_pretrained` are separate calls that don't read `model_init_kwargs`) or the reward model and reward tokenizer loads in GRPO / RLOO / self-distillation

Forcing users to thread `trust_remote_code=True` through several independent kwarg dicts is error-prone. Having a single arg is nicer.

### `ModelConfig.trust_remote_code` removed

`ModelConfig` already had a `trust_remote_code` field, so combining it with a trainer config in `TrlParser` would register `--trust_remote_code` twice and crash at parser construction.

Since the trainer-config flag now owns this, `ModelConfig.trust_remote_code` is removed:
- when possible, we rewire from `model_args.trust_remote_code` to `training_args.trust_remote_code`.
- when not possible, we don't rewire, (in example), which is not breaking

CLI behaviour is unchanged: `--trust_remote_code` now binds to `training_args.trust_remote_code`.

Related: https://github.com/huggingface/trl/pull/4186

---

> [!NOTE]
> **Medium Risk**
> Changes how models and tokenizers load across many trainers and touches the security-sensitive remote-code path, though the default remains False and behavior is more consistent than before.
> 
> **Overview**
> Adds a **`trust_remote_code`** flag (default `False`) on trainer configs and wires it through model, tokenizer/processor, reference, reward, and teacher loads so one CLI/config switch covers paths that `model_init_kwargs` alone did not.
> 
> **`ModelConfig.trust_remote_code` is removed** to avoid duplicate `--trust_remote_code` when parsing combined dataclasses; official `trl/scripts/*` now pass `training_args.trust_remote_code` into `model_init_kwargs`. Example scripts that pre-load models no longer thread the flag from `ModelConfig`.
> 
> **`create_model_from_path`** gains fallback logic for Hub remote-code checkpoints via `config.auto_map`. A **tiny `RemoteForCausalLM`** generator and **`test_trust_remote_code`** regressions assert trainers reject custom-code models without the flag and load them when it is set.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit a5d00a7c41f8b61a55634b2e2e530dfc3a8a9e20. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: examples/scripts/ppo/ppo.py [hunk 1]
```diff
@@ -86,31 +86,19 @@
         model_kwargs["device_map"] = get_kbit_device_map()
         model_kwargs["quantization_config"] = quantization_config
 
-    tokenizer = AutoTokenizer.from_pretrained(
-        model_args.model_name_or_path, padding_side="left", trust_remote_code=model_args.trust_remote_code
-    )
+    tokenizer = AutoTokenizer.from_pretrained(model_args.model_name_or_path, padding_side="left")
```
Comment (qgallouedec) [inline examples/scripts/ppo/ppo.py:89]: we don't support `trust_remote_code` in the examples. But these are "examples" only so it not to considered as breaking

File: examples/scripts/bco.py [hunk 2]
```diff
@@ -111,16 +111,10 @@ def mean_pooling(model_output, attention_mask):
...
-    )
-    ref_model = AutoModelForCausalLM.from_pretrained(
-        model_args.model_name_or_path, trust_remote_code=model_args.trust_remote_code
-    )
+    model = AutoModelForCausalLM.from_pretrained(model_args.model_name_or_path)
+    ref_model = AutoModelForCausalLM.from_pretrained(model_args.model_name_or_path)
 
-    tokenizer = AutoTokenizer.from_pretrained(
-        model_args.model_name_or_path, trust_remote_code=model_args.trust_remote_code
-    )
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply examples/scripts/bco.py:123]: that's fine, the user will have to modify the script

File: trl/trainer/sft_config.py [hunk 3]
```diff
@@ -149,6 +153,13 @@ class SFTConfig(_BaseConfig):
             "in this dictionary."
         },
     )
+    trust_remote_code: bool = field(
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/sft_config.py:164]: fixed

File: trl/experimental/distillation/distillation.py [hunk 4]
```diff
@@ -107,7 +106,6 @@ def main(script_args, training_args, model_args):
 
     teacher_model_kwargs = dict(
         revision=training_args.teacher_model_revision,
-        trust_remote_code=model_args.trust_remote_code,
```
[the remark this answers is not in the document]
Comment (albertvillanova) [reply trl/experimental/distillation/distillation.py:110]: I think this makes sense.

**Review by albertvillanova — CHANGES_REQUESTED** [review]
Thanks for the improvement.

I noticed you also applied the change to the experimental trainers (KTO, SSD, TPO,...), which is great. However, the implementation seems only partial there: I think a few experimental trainers were missed.

**Review by kashif — APPROVED** [review]
